compiler/rustc_ast_lowering/src/expr.rs

```diff
@@ -410,7 +436,39 @@ impl<'hir> LoweringContext<'_, 'hir> {
                     self.create_def(parent_def_id, node_id, kw::Empty, DefKind::AnonConst, f.span);
                 }
 
-                let anon_const = AnonConst { id: node_id, value: arg };
+                let mut visitor = WillCreateDefIdsVisitor {};
+                let const_value = if let ControlFlow::Break(span) = visitor.visit_expr(&arg) {
+                    let mut const_args = vec![];
+                    let mut other_args = vec![];
+                    for (idx, arg) in args.iter().enumerate() {
```

Comment:
Can you move all this logic outside of the outer for loop. i.e. having two top level `for (idx, arg) in args.iter()` loops in this function. One that just finds all the spans needed to emit an error and then the existing loop (pre this PR) that is responsible for actually lowering the function call.